### ISSUE TITLE
Feature request: Add feature flags for backends

### ISSUE THREAD

**Issue by nick42d** [ratatui-org/ratatui#1319]
## Problem
I would like to remove the explicit `crossterm` dependency from my crate and instead utilise the re-exported version from `ratatui` to avoid #1298 . However, I'd also like to use `crossterm`'s `event-stream` feature.

## Solution
Add new feature flags to enable the corresponding features in the backends e.g new feature `crossterm-event-stream` to enable `crossterm/event-stream`. This seems consistent with the approach to re-export the backends.

**Comment by joshka** [comment]
I wonder whether we should be passing `no-default-features` to crossterm / termion / termwiz (and only include the features that we 100% need). This might make it easier for widget crates to compile a bit more agnostic to what features are enabled, but might make it more difficult for apps. I think this might need some exploration to understand which is the better of the two trade offs, in particular the impact of doing this for windows (as crossterm has a windows feature flag).

**Comment by nick42d** [comment]
> I wonder whether we should be passing `no-default-features` to crossterm / termion / termwiz (and only include the features that we 100% need).

This kind of approach, to provide feature flags for backend features, could allow libraries to allow opt out of crossterm default features whilst letting ratatui retain the current defaults - e.g:

```
...
[dependencies]
crossterm = { version = "0.28.1", default-features = false, optional = true }
...
[features]
default = ["crossterm-default", "underline-color"]
crossterm = ["dep:crossterm"]
crossterm-default = ["crossterm", "crossterm/default"]
```

Wouldn't widget library authors be better off using ratatui itself without the `crossterm` feature though? (since `Widget` is backend agnostic)

**Comment by matta** [comment]
> I wonder whether we should be passing no-default-features to crossterm / termion / termwiz (and only include the features that we 100% need).

This idea feels right on the face of it, but I think you have stated the right set of pros/cons as well. Another to add: reducing the feature set used by ratatui would be another ratatui major version bump.

**Comment by matta** [comment]
> Add new feature flags to enable the corresponding features in the backends e.g new feature crossterm-event-stream to enable crossterm/event-stream. This seems consistent with the approach to re-export the backends.

I am sympathetic to the annoyance of #1298, but I think providing these kinds of "proxy" features is an unfortunate piece of additional complexity.

### TLDR

I'm personally against the idea, because I think the problem is relatively minor, is well understood in the rust ecosystem, would accrue in ratatui some technical debt without addressing the root problem. That said, I can understand going forward with it anyway based on a pragmatic argument alone.

### Perhaps pragmatic, but unfortunate

I agree that providing the "proxy" features is a pragmatic solution to a real problem. In fact, there is precedent for this in some crates. For example, axum apparently has an "http" feature that merely enables the feature of the same name in one of its dependencies. Presumably, though, axum maintainers found that this feature was such a common use case, and was so commonly requested among its users, that providing the proxy feature was worthwhile. I'm not sure if we're there with extra crossterm features.

Here I'm merely advancing arguments that adding these kinds of features so would be merely _unfortunate_, while perhaps also being pragmatic.

### Direct dependencies are, most often, the right way to express a dependency on crate features -- the problem is with versions, not features

To re-state the problem (and risk re-stating the obvious): [feature unification](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification) allows crates to add to the set of enabled features for creates they depend directly on. So, setting extra crossterm features can be achieved by depending on crossterm directly and setting the features. This leads to frustrations with version skew, or #1298, admittedly.

Here, I think, *the direct dependency is correct*. If create A needs feature X from crate B, then crate A really does have a direct dependency on B (otherwise it wouldn't care about the feature!). So, I don't see idioms that move away from expressing this with a direct dependency as clearly good ideas.

That said, the problem is that *keeping the versions of dependencies in sync is annoying* when using current rust tooling in straightforward and recommended ways.

The "proxy feature" idea doesn't address the root issue. Yes, they let us state dependencies fewer times in the crate graph, thus reducing the version skew problem, but at the cost of moving the crate graph *away* from the desirable situation where creates explicitly and clearly state what features they need of the crates they use.

### Reasons against

1. At least in the case of crossterm, major version bumps have been infrequent recently. My guess is that this will continue to be so. So, the version skew is likely to be occasionally annoying, but not a major and frequent problem for ratatui users.
2. Crossterm documentation will describe features with one name, and ratatui documentation will describe the exact same features with, presumably, a different name. A future reader must understand both ratatui and crossterm to understand how a create uses the particular crossterm feature. This is an _unfortunate_ complexity.
3. Providing "proxy features" for dependent crates ties the crate more tightly to the dependencies than simply re-exporting their API. Concretely, ratatui would have to expose proxy configs for options ratatui itself never uses. So, not only would ratatui depend on the features that it does use, but also all features it doesn't. The addition, or removal, of any feature from any of ratatui's dependent terminal crates is a coordinated dance between ratatui and that crate. This technical debt is unfortunate.
4. If every crate in ratatui's situation did this, the crate graph could become littered with a great many "proxy features", which would arguably be quite confusing and messy, and explode into a maintenance problem quite quickly.
5. If a create uses feature X from another crate, it seems quite reasonable for that create to depend directly on it and enable the X feature. Take the example of using crossterm's `event-stream` feature. Ratatui doesn't use it, so why is ratatui involved at all in the effort to enable it?

### Possibly better in the future?

[RFC 3616: Public/Private dependencies](https://rust-lang.github.io/rfcs/3516-public-private-dependencies.html) talks about addressing this at the `rustc/cargo` level by allowing creates to depend "publicly" or "privately" on a crate. If I understand the proposal, this allows tooling to pick compatible versions better, report incompatible versions in with better errors, allow for better semver checks at release time, and more.

Another smaller idea would be to allow one create to express that it depends on the "same" version of a crate that one of its direct dependencies does. Not sure if that would fly, though.

**Comment by joshka** [comment]
Thanks @matta - lots to think about

> If a create uses feature X from another crate, it seems quite reasonable for that create to depend directly on it and enable the X feature. Take the example of using crossterm's event-stream feature. Ratatui doesn't use it, so why is ratatui involved at all in the effort to enable it?

I think I'm most convinced by this point, but it's definitely one that I need to give some more thought.

**Comment by nick42d** [comment]
Thanks for the comprehensive response @matta, and I do agree that having to maintain someone else's feature flags does increase maintenance burden for ratatui, and I too hope cargo provides a better solution in future (although I question the term technical debt here - you're not exactly borrowing something you'd need to pay back later, you're simply increasing the scope of 'responsibilities' that ratatui would have).

To me there is a bit of tension here with the rerexports as it feels like you are saying 'ratatui can help you avoid #1298 by providing rerexports - but only if you are using the default features of crossterm'.

Wondering if some documentation could help here - current docs say to use `ratatui::crossterm` but don't mention the pros and cons of this over using crossterm directly.

**Comment by joshka** [comment]
Yeah - this definitely needs some more docs - 1298 was a placeholder for that somewhat.
I see the most benefit of the re-export as being something which can help library users more-so than app users, and I'm willing to walk back the app advice of using ratatui::crossterm if it makes things easier for users generally. But it really comes down to understanding each of the various problems (there's some tension between each approach for libs / apps)

**Comment by matta** [comment]
> To me there is a bit of tension here with the rerexports as it feels like you are saying 'ratatui can help you avoid https://github.com/ratatui-org/ratatui/issues/1298 by providing rerexports - but only if you are using the default features of crossterm'.

I did not intend to imply the "_but only if you are using the default features of crossterm_" part.

I think I meant to say that if, and only if, ratatui reduces the set of features it enables in crossterm, *then* more app crates might find reasons to enable crossterm features, and choose to depend on crossterm directly in order to do that, and risk running into the version skew problem.

> I see the most benefit of the re-export as being something which can help library users more-so than app users

@joshka, I'm not understanding you here.

In what scenarios are you thinking that a library crate would benefit more than an app crate from ratatui feature flag re-exports?

> I'm willing to walk back the app advice of using ratatui::crossterm if it makes things easier for users generally.

Are there any such scenarios? I'm not able to come up with one. The API exposed by `ratatui::crosstool` seems like a good thing across the board.

**Comment by matta** [comment]
I just did some experiments and was surprised at the result.

I ran through https://ratatui.rs/tutorials/hello-world/ with modifications:

I specified this in `Cargo.toml`:

```toml
[dependencies]
crossterm = "0.28.1"
ratatui = "0.27.0"
```

I referred to the top level `cargo` crate in `main.rs`:

```rust
use crossterm::{
    event::{self, KeyCode, KeyEventKind},
    terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
    ExecutableCommand,
};
use ratatui::{backend::CrosstermBackend, style::Stylize, widgets::Paragraph, Terminal};
```

This gave me a frankenstein program that was using two crossterm versions. I verified this by looking at `cargo tree` output. Ratatui depends on crossterm 0.27.0 and my hello-world crate depends directly on crossterm 0.28.1.

Surprising result: the program compiles and run fine!

I can only assume this works in part by accident of the crossterm implementation and API. The render-side of crossterm writes to stdout. The program's `main.rs` does this through some basic setup (e.g. `disable_raw_mode`), but these don't interact with ratatui's use of crossterm. Then my program *only* reads stdin using crossterm 0.28.1 code, and ratatui *only* writes to stdout using crossterm 0.27.0 code. It all works out fine!

I suspect that if `ratatui::backend::CrosstermBackend` worked differently, and required, say, a value of some crossterm type as an argument, passed from my `main.rs`, then I would run into problems here.

@nick42d I wonder if you could take your program, depend on crossterm directly, enable the crossterm/event-stream feature, and induce cargo to use a different version of crossterm this way, and have it all work out similarly well?

**Comment by matta** [comment]
Looks like https://github.com/EmbarkStudios/cargo-deny seems to be a popular way people maintain code bases with some handle on their duplicate dependencies.

The existence of `cargo-deny` and similar crates like https://github.com/Keruspe/cargo-duplicates suggest to me that this is an ecosystem problem with some commonly used mitigations.

**Comment by nick42d** [comment]
> @nick42d I wonder if you could take your program, depend on crossterm directly, enable the crossterm/event-stream feature, and induce cargo to use a different version of crossterm this way, and have it all work out similarly well?

Currently I will just make sure I depend on the same version of crossterm as ratatui. If ratatui provides an alternative in future I will migrate, so I will keep in eye on this issue.

**Comment by joshka** [comment]
To the best of my knowledge, we don't expose crossterm on our public API right now anywhere. Which means that using a different version may just work.

In the next point release of Ratatui, we're [introducing some new simple initialization code](https://github.com/ratatui-org/ratatui/pull/1289/). This will call `enable_raw_mode()`. This method stores the current termios in a static global field so that the terminal can be restored on application exit. Calling `disable_raw_mode()` from a different version of Crossterm will see a different global static field and won't work as expected.

There are other similar things which may cause problems if mixed versions are used. On the input side, the buffers that keep events when they're polled / read are likely version specific in the same way (due to global state). I'm unsure if any of the other methods that we call like ones to do with window / terminal size have any interactions with global state.

TL;DR - the mixed versions risks are likely only going to be seen as behavior problems due to global state.

Looking to the future, I wouldn't discount widget crates / other crates wanting to use the events system generally and hence running into this sort of problem if they don't use the same version of Crossterm as other libs and the app. I also would not be surprised to find other non-trivial incompatibilities show up.

**Comment by orhun** [comment]
Now that we have `init()` and `restore()` methods (#1289), it might be a good idea to experiment with mixed versions / imports again and update the crate documentation with use-cases/pros/cons.
